pgrx/src/memcx.rs

```diff
@@ -45,3 +55,99 @@ where
 
     f(&memcx)
 }
+
+/// Does alignment / padding logic for pre-Postgres16 8-byte alignment.
+#[cfg(all(
+    feature = "nightly",
+    any(feature = "pg12", feature = "pg13", feature = "pg14", feature = "pg15")
+))]
+fn layout_for_pre16(
+    layout: std::alloc::Layout,
+) -> Result<std::alloc::Layout, std::alloc::AllocError> {
+    Ok(layout.align_to(8).map_err(|_e| std::alloc::AllocError)?.pad_to_align())
+}
+
+#[cfg(feature = "nightly")]
+unsafe impl<'mcx> std::alloc::Allocator for &MemCx<'mcx> {
+    // Future-proofing - currently only pg16 supports arbitrary alignment, but that is likely
+    // to change, whereas old versions are unlikely to lose 8-byte alignment.
+    #[cfg(any(feature = "pg12", feature = "pg13", feature = "pg14", feature = "pg15"))]
+    fn allocate(
+        &self,
+        layout: std::alloc::Layout,
+    ) -> Result<NonNull<[u8]>, std::alloc::AllocError> {
+        // On versions before Postgres 16, alignment is always 8 byte / 64 bit.
+        let layout = layout_for_pre16(layout)?;
+        unsafe {
+            let ptr = pgrx_pg_sys::MemoryContextAlloc(self.ptr.as_ptr(), layout.size());
+            let slice: &mut [u8] = slice::from_raw_parts_mut(ptr.cast(), layout.size());
+            Ok(NonNull::new_unchecked(slice))
+        }
+    }
```

Comment:
This is not correct. You cannot achieve the goal you are trying to accomplish by tinkering with `alloc::Layout`. The `Layout` is a *command* to you. You obey or you fail, you must not modify the `Layout`. Please just return an AllocError always on these versions for now.

Reply:
Should be corrected in https://github.com/pgcentralfoundation/pgrx/pull/1737/commits/f149072c318f7e39c212daf3b59e0f723d1dfd46